src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+                IClock clock)
+            {
+                _awsCredentials = Ensure.IsNotNull(awsCredentials, nameof(awsCredentials));
+                _randomByteGenerator = Ensure.IsNotNull(randomByteGenerator, nameof(randomByteGenerator));
+                _clock = Ensure.IsNotNull(clock, nameof(clock));
+            }
+
+            public string Name
+            {
+                get { return MechanismName; }
+            }
+
+            public ISaslStep Initialize(IConnection connection, SaslConversation conversation, ConnectionDescription description)
+            {
+                Ensure.IsNotNull(connection, nameof(connection));
+                Ensure.IsNotNull(description, nameof(description));
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument
+                {
+                    { "r", nonce },
+                    { "p", (int)'n' }
+                };
+
+                var clientMessageBytes = document.ToBson();
+
+                return new ClientFirst(clientMessageBytes, nonce, _awsCredentials, _clock);
+            }
+
+            private byte[] GenerateRandomBytes()
+            {
+                return _randomByteGenerator.Generate(ClientNonceLength);
+            }
+        }
+
+        private class ClientFirst : ISaslStep
+        {
+            private readonly AwsCredentials _awsCredentials;
+            private readonly byte[] _bytesToSendToServer;
+            private readonly IClock _clock;
+            private readonly byte[] _nonce;
+
+            public ClientFirst(
+                byte[] bytesToSendToServer,
+                byte[] nonce,
+                AwsCredentials awsCredentials,
+                IClock clock)
+            {
+                _bytesToSendToServer = bytesToSendToServer;
+                _nonce = nonce;
+                _awsCredentials = awsCredentials;
+                _clock = clock;
+            }
+
+            public byte[] BytesToSendToServer
+            {
+                get { return _bytesToSendToServer; }
+            }
+
+            public bool IsComplete
+            {
+                get { return false; }
+            }
+
+            public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedFromServer)
+            {
+                var serverFirstMessageDocument = BsonSerializer.Deserialize<BsonDocument>(bytesReceivedFromServer);
+                var serverNonce = serverFirstMessageDocument["s"].AsByteArray;
+                var host = serverFirstMessageDocument["h"].AsString;
+
+                if (serverNonce.Length != ClientNonceLength * 2 || !serverNonce.Take(ClientNonceLength).SequenceEqual(_nonce))
+                {
+                    throw new MongoAuthenticationException(conversation.ConnectionId, "Server sent an invalid nonce.");
+                }
+                if (host.Length < 1 || host.Length > 255 || host.Contains(".."))
```

Comment:
It sounds like we don't have tests for this case (and I guess some other). But I will rely on @jyemin who thinks that we don't need them

Reply:
The spec says:
>Drivers MUST reject FQDN names with empty labels, e.g., "abc..def", and error on any additional fields

I'm not sure that I understand this part: `and error on any additional fields`.
@MikalaiMazurenka  have you discussed this part?

Reply:
Thanks for notice of `and error on any additional fields`!
Also, I've added a few tests.
Done.